packages/firestore/src/index/ordered_code_writer.ts

```diff
@@ -0,0 +1,162 @@
+ */
+const DEFAULT_BUFFER_SIZE = 1024;
+
+/**
+ * Converts a JavaScript number to a byte array (using little endian
+ * encoding).
+ */
+function doubleToLongBits(value: number): Uint8Array {
+  const dv = new DataView(new ArrayBuffer(8));
+  dv.setFloat64(0, value, false);
+  return new Uint8Array(dv.buffer);
+}
+
+/**
+ * Counts the number of zeros in a byte.
+ *
+ * Visible for testing.
+ */
+export function numberOfLeadingZerosInByte(x: number): number {
+  if (x === 0) {
+    return 8;
+  }
+
+  let zeros = 0;
+  if (x >> 4 === 0) {
+    // Test if the first four bits are zero.
+    zeros += 4;
+    x = x << 4;
+  }
+  if (x >> 6 === 0) {
+    // Test if the first two (or next two) bits are zero.
+    zeros += 2;
+    x = x << 2;
+  }
+  if (x >> 7 === 0) {
+    // Test if the remaining bit is zero.
+    zeros += 1;
+  }
+  return zeros;
+}
+
+/** Counts the number of leading zeros in the given byte array. */
+function numberOfLeadingZeros(bytes: Uint8Array): number {
+  let leadingZeros = 0;
+  for (let i = 0; i < bytes.length; ++i) {
+    const zeros = numberOfLeadingZerosInByte(bytes[i] & 0xff);
+    leadingZeros += zeros;
+    if (zeros !== 8) {
+      break;
+    }
+  }
+  return leadingZeros;
+}
+
+/**
+ * Returns the number of bytes required to store "value". Leading zero bytes
+ * are skipped.
+ */
+function unsignedNumLength(value: Uint8Array): number {
+  // This is just the number of bytes for the unsigned representation of the number.
+  const numBits = LONG_SIZE - numberOfLeadingZeros(value);
+  return Math.ceil(numBits / BYTE_SIZE);
+}
+
+/**
+ * OrderedCodeWriter is a minimal-allocation implementation of the writing
+ * behavior defined by the backend.
+ *
+ * The code is ported from its Java counterpart.
+ */
+export class OrderedCodeWriter {
+  buffer = new Uint8Array(DEFAULT_BUFFER_SIZE);
+  position = 0;
+
+  writeNumberAscending(val: number): void {
+    const value = this.toOrderedBits(val);
+    const len = unsignedNumLength(value);
+    this.ensureAvailable(1 + len);
+    this.buffer[this.position++] = len & 0xff; // Write the length
+    for (let i = value.length - len; i < value.length; ++i) {
+      this.buffer[this.position++] = value[i] & 0xff;
+    }
+  }
+
+  writeNumberDescending(val: number): void {
+    const value = this.toOrderedBits(val);
+    const len = unsignedNumLength(value);
+    this.ensureAvailable(1 + len);
+    this.buffer[this.position++] = ~(len & 0xff); // Write the length
+    for (let i = value.length - len; i < value.length; ++i) {
+      this.buffer[this.position++] = ~(value[i] & 0xff);
+    }
+  }
+
+  /**
+   * Encodes `val` into an encoding so that the order matches the IEEE 754
+   * floating-point comparison results with the following exceptions:
+   *   -0.0 < 0.0
+   *   all non-NaN < NaN
+   *   NaN = NaN
+   */
+  private toOrderedBits(val: number): Uint8Array {
+    const value = doubleToLongBits(val);
+    const isNegative = (value[0] & 0x80) !== 0;
```

Comment:
I'm curious why `const isNegative = val < 0` is not used? Is it because of the three exceptions listed above?

Reply:
We are not actually looking at a number that is encoded as a long. Instead, we are looking at the first element of the byte array that is encoded as a number from 0 to 255.

Reply:
that's my question.. wouldn't it be easier to detect negativeness by checking the number (`val`) rather than checking the byte array (`value`)?

No action is needed though.

Reply:
We are checking whether the bit representation is "negative" not the number itself. Maybe the naming could use some improvement, but this is what the original implementation uses.

The tests fail if we just check if `val` is negative.